### ISSUE TITLE
Chef.log_deprecation triggers infinite recursion.

### ISSUE THREAD

**Issue by coderanger** [chef/chef#3969]
Adding this for tracking, the Poise nightly build died so one of the commits in the last 24 hours changed something, might be a reversion worth fixing or might be a private API I shouldn't be using. /cc @jkeiser

**Comment by coderanger** [comment]
https://travis-ci.org/poise/poise/jobs/82239498 is the broken build, looks like all the failures are an infinite recursion issue related to the DSL.

**Comment by coderanger** [comment]
The line triggering the stack overflow is:

```ruby
sub_run_context.resource_collection = Poise::Subcontext::ResourceCollection.new(parent_context.resource_collection)
```

Though I can repro with just `sub_run_context.resource_collection = nil` too.

**Comment by coderanger** [comment]
`RunContext.resource_collection=` is still (incorrectly but that's a different bug) logging a deprecation warning so I'm betting it is related to that.

**Comment by coderanger** [comment]
And the trace is:

```
lib/chef/run_context.rb:535 Chef::RunContext::Deprecated#resource_collection=
lib/chef/chef_class.rb:207 Chef#log_deprecation
lib/chef/event_dispatch/dispatcher.rb:49 Chef::EventDispatch::Dispatcher#deprecation
lib/chef/event_dispatch/dispatcher.rb:28 Chef::EventDispatch::Dispatcher#call_subscribers
lib/chef/formatters/base.rb:215 Chef::Formatters::Base#deprecation
lib/chef/chef_class.rb:207 Chef#log_deprecation
lib/chef/event_dispatch/dispatcher.rb:49 Chef::EventDispatch::Dispatcher#deprecation
lib/chef/event_dispatch/dispatcher.rb:28 Chef::EventDispatch::Dispatcher#call_subscribers
lib/chef/formatters/base.rb:215 Chef::Formatters::Base#deprecation
lib/chef/chef_class.rb:207 Chef#log_deprecation
lib/chef/event_dispatch/dispatcher.rb:49 Chef::EventDispatch::Dispatcher#deprecation
lib/chef/event_dispatch/dispatcher.rb:28 Chef::EventDispatch::Dispatcher#call_subscribers
```

**Comment by coderanger** [comment]
Fixed via #3971
